plugins/datastreamer_plugin/utils/dynamic_types_utils.cpp

```diff
@@ -34,149 +35,260 @@ namespace eprosima {
-            case eprosima::fastrtps::types::TK_MAP:
-                // Complex types
-                // TODO
-                WARNING("DataType member " << members.first << "with not supported kind " << kind);
-                break;
-
-            default:
-                // Shouold not happen
-                throw InconsistencyException("Kind not in switch");
+            // Using the Dynamic type, retrieve the name of the fields and its descriptions
+            std::map<std::string, DynamicTypeMember*> members_by_name;
+            type->get_all_members_by_name(members_by_name);
+
+            for (auto& members : members_by_name)
+            {
+                std::vector<MemberId> new_members_tree(current_members_tree);
+                new_members_tree.push_back(members.second->get_id());
+                std::vector<TypeKind> new_kinds_tree(current_kinds_tree);
+                new_kinds_tree.push_back(kind);
+
+                get_introspection_type_names(
+                    base_type_name + separator + members.first,
+                    members.second->get_descriptor()->get_type(),
+                    numeric_type_names,
+                    string_type_names,
+                    new_members_tree,
+                    current_kinds_tree,
+                    separator);
+            }
+            break;
         }
+
+        case TK_BITSET:
+        case TK_UNION:
+        case TK_SEQUENCE:
+        case TK_MAP:
+        case TK_NONE:
+        default:
+            WARNING(kind << " DataKind Not supported");
+            throw std::runtime_error("Unsupported Dynamic Types kind");
     }
 }
 
-void get_introspection_data(
-        eprosima::fastrtps::types::DynamicType_ptr type,
+void get_introspection_numeric_data(
         const TypeIntrospectionCollection& numeric_type_names,
-        const TypeIntrospectionCollection& string_type_names,
-        eprosima::fastrtps::types::DynamicData_ptr data,
-        TypeIntrospectionNumericDatum& numeric_data,
-        TypeIntrospectionStringDatum& string_data)
+        const DynamicData_ptr& data,
+        TypeIntrospectionNumericStruct& numeric_data_result)
 {
-    // First get numeric data
-    // It is used by index of the vector to avoid unnecessary lookups, but vectors must be sorted
+    DEBUG("Getting numeric data");
+
+    // It is used by index of the vector to avoid unnecessary lookups.
+    // Vectors must be sorted in the same order => creation order given in get_introspection_type_names
     for (int i = 0; i < numeric_type_names.size(); ++i)
     {
-        numeric_data[i].second = get_numeric_type_from_data(
+        const auto& member_type_info = numeric_type_names[i];
+
+        // Get reference to variables to avoid calling get twice
+        const std::vector<MemberId>& members =
+            std::get<std::vector<MemberId>>(member_type_info);
+        const std::vector<TypeKind>& kinds =
+            std::get<std::vector<TypeKind>>(member_type_info);
```

Comment:
typical use case for `const auto& members = ...`, for less verbosity (type duplicated in the expression)